# First print your name in the cell below then save this file. (or something nice about your instructor)

In [ ]:
print('Brendan Costello')

# Data wrangling with Pandas exercise
* For this exercise we will be using the `listings.csv` data file.

In [2]:
import pandas as pd
import numpy as np

# Load in the data file using `pd.read_csv()` and take note of column names, shape, and dtypes

In [3]:
listings = pd.read_csv('/Users/brendancostello/Desktop/Brendan/CTP/2023-fall-data-science-fridays/Week-1-Pandas/data/listings.csv')

In [4]:
listings.dtypes

id                                  int64
name                               object
host_id                             int64
host_name                          object
neighbourhood_group                object
neighbourhood                      object
latitude                          float64
longitude                         float64
room_type                          object
price                               int64
minimum_nights                      int64
number_of_reviews                   int64
last_review                        object
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
dtype: object

In [5]:
listings.head()

id                                               name  host_id  \
0  2595                              Skylit Midtown Castle     2845   
1  3831  Whole flr w/private bdrm, bath & kitchen(pls r...     4869   
2  5121                                    BlissArtsSpace!     7356   
3  5136           Spacious Brooklyn Duplex, Patio + Garden     7378   
4  5178                   Large Furnished Room Near B'way　     8967   

     host_name neighbourhood_group       neighbourhood  latitude  longitude  \
0     Jennifer           Manhattan             Midtown  40.75362  -73.98377   
1  LisaRoxanne            Brooklyn        Clinton Hill  40.68514  -73.95976   
2        Garon            Brooklyn  Bedford-Stuyvesant  40.68688  -73.95596   
3      Rebecca            Brooklyn         Sunset Park  40.66120  -73.99423   
4     Shunichi           Manhattan      Hell's Kitchen  40.76489  -73.98493   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0  Entire home/apt    175               3                 48  2019-11-04   
1  Entire home/apt     75               1                340  2020-08-01   
2     Private room     60              29                 50  2019-12-02   
3  Entire home/apt    175              14                  1  2014-01-02   
4     Private room     65               2                473  2020-03-15   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.37                               2               365  
1               4.75                               1               265  
2               0.37                               1               365  
3               0.01                               1               295  
4               3.44                               1               340

In [6]:
listings.tail()

id                                               name    host_id  \
46522  44807522         Designer Gramercy Studio Townhouse by UNSQ   12941925   
46523  44807786  Cozy & comfy apt in the heart of Inwood Manhattan  284790520   
46524  44811717  Comfortable safe environment 24hr security camera  362453686   
46525  44814944                 Upper West Side studio 86th Street    4039777   
46526  44818009  5MIN D/N trains, NEAR THE BEACH, 50’ TO MANHATTAN   48098268   

      host_name neighbourhood_group       neighbourhood  latitude  longitude  \
46522     Brian           Manhattan            Gramercy  40.73433  -73.98383   
46523     Salar           Manhattan  Washington Heights  40.85820  -73.92733   
46524    Nicole            Brooklyn       East Flatbush  40.65399  -73.93287   
46525  Fernando           Manhattan     Upper West Side  40.78731  -73.97029   
46526    Marina            Brooklyn           Gravesend  40.59945  -73.98209   

             room_type  price  minimum_nights  number_of_reviews last_review  \
46522  Entire home/apt    145               7                  0         NaN   
46523  Entire home/apt     87               6                  0         NaN   
46524     Private room     59               3                  0         NaN   
46525  Entire home/apt     80              30                  0         NaN   
46526     Private room     66               1                  0         NaN   

       reviews_per_month  calculated_host_listings_count  availability_365  
46522                NaN                               1               164  
46523                NaN                               2                85  
46524                NaN                               1                90  
46525                NaN                               1               113  
46526                NaN                               1                38

In [8]:
listings.shape

(46527, 16)

## Exercise 2 - Filtering

Return the following subsets of the dataframe.

1. How many listings are there with a price less than 100? 


2. Find how many listings there are in just Brooklyn.


3. Find how many listings there are in Brooklyn with a price less than 100.


4. Using `.isin()` select anyone that has the host name of Michael, David, John, and Daniel.


5. Create a new column called `adjusted_price` that has $100 added to every listing in Williamsburg.  The prices for all other listings should be the same as the were before. 


6. What % of the rooms are private, and what % of the rooms are shared.  
    * Hint, use `.value_counts()`


# 1. How many listings are there with a price less than 100? 

In [9]:
# Create the variable housing the information requested
houses_under_100 = listings[listings['price'] < 100]

# Call the variable shape to see the information
houses_under_100.shape

(22778, 16)

### **Answer**- There are 22,778 rows representing listings for less than $100 in this dataframe

# 2. Make a new DataFrame of listings in Brooklyn named `df_bk` and find how many listings in just Brooklyn.



In [10]:
# Create the variable housing the information requested
df_bk = listings[listings['neighbourhood_group'] == 'Brooklyn']

# Call the variable shape to see the information
df_bk.shape

(18632, 16)

### **Answer-** There are 18,632 rows representing Brooklyn in this dataframe

# 3. Find how many listings there are in Brooklyn with a price less than 100.

In [11]:
# Create the variable housing the information requested
df_bk_cheap = df_bk[df_bk['price'] < 100]

# Call the variable shape to see the information
df_bk_cheap.shape

(10473, 16)

### **Answer-** There are 10,473 rows in Brooklyn with a price less than $100 in this dataframe

# 4. Using `.isin()` select anyone that has the host name of Michael, David, John, and Daniel. How many total are there that have those names



In [12]:
# Create the variable housing the information requested INCLUDING the .isin() function
hosts = listings[listings["host_name"].isin(['Michael', 'David', 'John', 'Daniel'])]

# Call the variable shape to see the information
hosts.shape

(1258, 16)

### **Answer-** There are 1,258 hosts with one of those names

# 5. Create a new column called `adjusted_price` that has $100 added to every listing in Williamsburg. The prices for all other listings should be the same as the were before. 



In [13]:
# Create a condition where 'neighbourhood' == 'Williamsburg'
condition = df_bk['neighbourhood'] == 'Williamsburg'

# Create a second condition where the figure in 'price' has 100 added to it
adj_price = df_bk['price'] + 100

# Create a column called 'adjusted_price and use numpy to add the second condition to rows that meet the first one
df_bk['adjusted_price'] = np.where(condition, adj_price, 0)

print(df_bk[condition])

             id                                               name    host_id  \
11         6848                   Only 2 stops to Manhattan studio      15991   
15         7801                   Sweet and Spacious Brooklyn Loft      21207   
35        14290                         * ORIGINAL BROOKLYN LOFT *      56104   
38        14377                   Williamsburg 1 bedroom Apartment      56512   
42        15385                              Very, very cozy place      60252   
...         ...                                                ...        ...   
46484  44746041      Modern, light-filled apartment in Wiliamsburg    3317380   
46489  44749327  Luxury 3 bedroom apartment Williamsburg ROOF P...  334780877   
46493  44753863            Bright, spacious East Williamsburg loft   38062951   
46513  44800276        Williamsburg 1 Bedroom, Spacious and Classy  361034257   
46517  44802032                  A very large room in Williamsburg  213793805   

             host_name neig

/var/folders/gf/0_8jxzcx2p54tfxnbprqq9xr0000gn/T/ipykernel_47232/3839607196.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_bk['adjusted_price'] = np.where(condition, adj_price, 0)


In [20]:
df_bk.head()

id                                               name  host_id  \
1   3831  Whole flr w/private bdrm, bath & kitchen(pls r...     4869   
2   5121                                    BlissArtsSpace!     7356   
3   5136           Spacious Brooklyn Duplex, Patio + Garden     7378   
8   5803     Lovely Room 1, Garden, Best Area, Legal rental     9744   
11  6848                   Only 2 stops to Manhattan studio    15991   

        host_name neighbourhood_group       neighbourhood  latitude  \
1     LisaRoxanne            Brooklyn        Clinton Hill  40.68514   
2           Garon            Brooklyn  Bedford-Stuyvesant  40.68688   
3         Rebecca            Brooklyn         Sunset Park  40.66120   
8          Laurie            Brooklyn         South Slope  40.66829   
11  Allen & Irina            Brooklyn        Williamsburg  40.70837   

    longitude        room_type  price  minimum_nights  number_of_reviews  \
1   -73.95976  Entire home/apt     75               1                340   
2   -73.95596     Private room     60              29                 50   
3   -73.99423  Entire home/apt    175              14                  1   
8   -73.98779     Private room     88               4                180   
11  -73.95352  Entire home/apt    109               2                181   

   last_review  reviews_per_month  calculated_host_listings_count  \
1   2020-08-01               4.75                               1   
2   2019-12-02               0.37                               1   
3   2014-01-02               0.01                               1   
8   2020-03-18               1.31                               3   
11  2020-03-16               1.32                               1   

    availability_365  adjusted_price  
1                265               0  
2                365               0  
3                295               0  
8                344               0  
11                 0             209

# 6. What % of the rooms are private, and what % of the rooms are shared.  

In [16]:
# Establish a variable for private rooms
private = df_bk['room_type'] == 'Private room'

# Establish a variable for all other rooms
shared = df_bk['room_type'] != 'Private room'

# Divide the variable for all other rooms by the total of all rooms
percent_priv = len(df_bk[shared])/len(df_bk)

# Divide the variable for all private rooms by the total of all rooms
percent_not_priv = len(df_bk[private])/len(df_bk)

# Print the results in the form of a percentage rounded to two decimal places
print('%.2f' % percent_priv, '% are Private')
print('%.2f' % percent_not_priv, '% are NOT Private')

0.51 % are Private
0.49 % are NOT Private


### **Answer-** 51% of rooms are Private, the other 49% are not private. NOTE: I could not figure out how to utilize .value_counts() as per the hint. After looking up the function and its purpose, I think I could put this in bins and get an accurate count as to how many rooms there are then divide by total rooms, but this method seems a lot easier.

# Exercise 3 - Grouping

1. Using `groupby`, count how many listings are in each neighbourhood_group.


2. Using `groupby`, find the mean price for each of the neighbourhood_groups. 


3. Using `groupby` and `.agg()`, find the min and max price for each of the neighbourhood_groups. 


4. Using `groupby`, find the median price for each room type in each neighbourhood_group.


5. Using `groupby` and `.agg()`, find the count, min, max, mean, median, and std of the prices for each room type in each neighbourhood_group.

# 1. Using `groupby`, count how many listings are in each neighbourhood_group.




In [21]:
neighborhood = listings.groupby('neighbourhood_group')
print(neighborhood['calculated_host_listings_count'].sum())

neighbourhood_group
Bronx              2619
Brooklyn          45790
Manhattan        180430
Queens            44534
Staten Island       692
Name: calculated_host_listings_count, dtype: int64


# 2. Using `groupby`, find the mean price for each of the neighbourhood_groups. 




In [22]:
print(neighborhood['price'].mean())

neighbourhood_group
Bronx             92.751479
Brooklyn         120.225258
Manhattan        191.880466
Queens            99.754965
Staten Island    110.947214
Name: price, dtype: float64


# 2.5. Using `groupby`, find the mean price for each room_type. 


In [23]:
room_types = listings.groupby('room_type')
print(room_types['calculated_host_listings_count'].sum())

room_type
Entire home/apt    165577
Hotel room           5105
Private room        99310
Shared room          4073
Name: calculated_host_listings_count, dtype: int64


In [24]:
print(room_types['price'].mean(), ['room-type'])

room_type
Entire home/apt    199.395950
Hotel room         275.015075
Private room        91.453084
Shared room         87.063830
Name: price, dtype: float64 ['room-type']


# 3. Using `groupby` and `.agg()`, find the min and max price for each of the neighbourhood_groups. 




In [25]:
neighborhood['price'].agg([ 'count', 'min', 'max'])

count  min    max
neighbourhood_group                   
Bronx                 1183   16   1404
Brooklyn             18632    0  10000
Manhattan            20580    0  10000
Queens                5791    0  10000
Staten Island          341   19   1200

# 4. Using `groupby`, find the mean price for each room_type in each neighbourhood_group.



In [26]:
room_df_mean1 = neighborhood['price'].agg(['mean'])

extracted_column1 = listings['room_type']
extracted_column2 = listings['price']

room_df_mean1.insert(1, 'room_type', extracted_column1)
room_df_mean1.insert(1, 'price', extracted_column2)

result = room_df_mean1.groupby(['neighbourhood_group', 'room_type'])['price'].mean().reset_index()

print(result)

Empty DataFrame
Columns: [neighbourhood_group, room_type, price]
Index: []


In [27]:
room_df_mean2 = neighborhood['price'].agg(['mean'])

extracted_column1 = listings['room_type']
extracted_column2 = listings['price']

room_df_mean2 = room_df_mean2.join(extracted_column1)
room_df_mean2 = room_df_mean2.join(extracted_column2)

result = room_df_mean2.groupby(['neighbourhood_group', 'room_type'])['price'].mean().reset_index()

print(result)

Empty DataFrame
Columns: [neighbourhood_group, room_type, price]
Index: []


# 5. Using `groupby` and `.agg()`, find the count, min, max, mean, median, and std of the prices for each room type in each neighbourhood_group.





In [28]:
neighborhood['price'].agg( ['count', 'min', 'max', 'mean', 'median', 'std', 'sum'] )

count  min    max        mean  median         std  \
neighbourhood_group                                                      
Bronx                 1183   16   1404   92.751479    69.0   95.787081   
Brooklyn             18632    0  10000  120.225258    90.0  180.136376   
Manhattan            20580    0  10000  191.880466   129.0  428.682578   
Queens                5791    0  10000   99.754965    70.0  207.648739   
Staten Island          341   19   1200  110.947214    79.0  122.676152   

                         sum  
neighbourhood_group           
Bronx                 109725  
Brooklyn             2240037  
Manhattan            3948900  
Queens                577681  
Staten Island          37833

# Join and file saving.
1. Load the `prices.csv` and the `n_listings.csv`


2. Do join that keeps all the records for each table.
    * Neighbourhood groups should include ['Bronx', 'Brooklyn', 'Manhattan', 'Queens', 'Staten Island',
       'LongIsland']
       
       
3. Save your joined csv as `joined.csv`


4. Load your saved table and see if it looks the same or different that the DataFrame you used to create it. 

# 1. Load the `prices.csv` and the `n_listings.csv`



In [29]:
prices = pd.read_csv('/Users/brendancostello/Desktop/Brendan/CTP/2023-fall-data-science-fridays/Week-1-Pandas/data/prices.csv')

n_listings = pd.read_csv('/Users/brendancostello/Desktop/Brendan/CTP/2023-fall-data-science-fridays/Week-1-Pandas/data/n_listings.csv')

In [30]:
prices.head()

neighbourhood_group   mean_price
0               Bronx    92.751479
1            Brooklyn   120.225258
2           Manhattan   191.880466
3              Queens    99.754965
4       Staten Island   110.947214

In [31]:
n_listings.head()

neighbourhood_group;n_listings
0                     Bronx;1183
1                 Brooklyn;18632
2                Manhattan;20580
3                LongIsland;4121

### **NOTE-** It took me an embarassing amount of time to remember to adjust the delimiter in order to get the answer to problem 2. It kept throwing the error 'neighbourhood_group' with no further explanation, so I thought it was a spelling error...which technically it was.

In [34]:
prices = pd.read_csv('/Users/brendancostello/Desktop/Brendan/CTP/2023-fall-data-science-fridays/Week-1-Pandas/data/prices.csv', sep=',')

n_listings = pd.read_csv('/Users/brendancostello/Desktop/Brendan/CTP/2023-fall-data-science-fridays/Week-1-Pandas/data/n_listings.csv', sep=';')

In [33]:
prices.head()

neighbourhood_group   mean_price
0               Bronx    92.751479
1            Brooklyn   120.225258
2           Manhattan   191.880466
3              Queens    99.754965
4       Staten Island   110.947214

In [35]:
n_listings.head()

neighbourhood_group  n_listings
0               Bronx        1183
1            Brooklyn       18632
2           Manhattan       20580
3          LongIsland        4121

# 2. Do join that keeps all the records for each table.




In [36]:
df = pd.merge(prices, n_listings, on='neighbourhood_group', how='outer')
df

neighbourhood_group   mean_price  n_listings
0               Bronx    92.751479      1183.0
1            Brooklyn   120.225258     18632.0
2           Manhattan   191.880466     20580.0
3              Queens    99.754965         NaN
4       Staten Island   110.947214         NaN
5          LongIsland          NaN      4121.0

## Assuming 'keeping all records' means an outer join since an inner join would lose rows 3, 4, and 5; a left join would lose rows 3 and 4; and a right join would lose row 5.

# Extra Credit
Every question below this cell is extra credit and optional.

### 1. (Easy) Explore this new PandasAI Package and tell me what its all about because I've never used it. 
* https://www.youtube.com/watch?v=5w6eZaoDVVk&ab_channel=CodingIsFun  
* See if you can use it on the listings.csv to find out some cool info. or answer some of the questions above. 

### 2. (Very Easy) Find other cool Panda packages / add ons and show us what they can do well. And how you installed them. 

### 3. (Medium) Use the grammys.csv data for the next section of questions.

1. Who was won Album of the Year in 2016?


2. Who won Best Rap Album in 2009?


3. How many awards was Kendrick Lamar nomiated for, and how many did he win...?

### (Hard) Load the Game Logs for 2022 and add the column names using a dictionary.  
* [Link to the data page](https://www.retrosheet.org/gamelogs/)
* [Link to the column names](https://procatinator.com/)
* haha, gotta find them yourself!

### (Extra Hard) Download the files for the past 5 years into a new folder and add them all into one data frame using pandas, then save that new file.
* Try to not hard code in the file names. We want to do this programmatically because what if we want to add new/more file names in the future.